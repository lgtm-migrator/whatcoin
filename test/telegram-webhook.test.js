/* eslint-disable jest/valid-title */
const telegramWebhook = require('../telegram-webhook');
const ctx = require('./defaultContext');
const nock = require('nock');
const filenamify = require('filenamify');

afterAll(() => nock.restore());

nock.back.fixtures = __dirname + '/__fixtures__/';
nock.back.setMode(process.env.CI ? 'lockdown' : 'record');

const msgReply = async (text, locale) => {
  const req = {
    body: {
      update_id: 1,
      message: { text, from: { language_code: locale }, chat: { id: 2 } },
    },
  };
  const res = await telegramWebhook(ctx, req);
  return res?.body;
};

const msgReplyText = async (text) => (await msgReply(text))?.text;

beforeEach(async () => {
  const state = expect.getState();
  state.nockBack = await nock.back(`${filenamify(state.currentTestName)}.json`);
});

afterEach(() => {
  const { nockBack } = expect.getState();
  nockBack.nockDone();
  nockBack.context.assertScopesFinished();
});

describe('webhook', () => {
  it('ignores everything except known commands', async () => {
    expect(await msgReplyText('/foo')).toBeUndefined();
    expect(await msgReplyText('bar')).toBeUndefined();
    expect(await msgReplyText()).toBeUndefined();
  });
});

describe('/version', () => {
  it('- shows name and version info', () =>
    expect(msgReply('/version')).resolves.toMatchObject({
      chat_id: 2,
      method: 'sendMessage',
      text: /Whatcoin v[\d.]+/,
    }));
});

const commandTests = {
  '/start': 'shows a welcome message',
  '/price': [
    ['defaults to btc in usd'],
    ['supports other currencies', 'ethereum eur'],
    ['finds coins by symbol', 'ETH GBP'],
    ['finds coins by name', 'PEAKDEFI'],
    ['finds coins by id partial match', 'bitm'],
    ['finds coins by symbol partial match', 'algoha'],
    ['finds coins by name partial match', 'peakd'],
    ['returns an error for invalid coin', 'asdfasdf eur'],
    ['returns an error for invalid vs currency', 'bitcoin asdasd'],
    ['redirects to convert when first arg is a number', '10000 doge'],
    ['redirects to convert when second arg is a number', 'doge 10000'],
  ],
  '/convert': [
    ['defaults to 1 usd in usd'],
    ['defaults to converting coins to usd', '42 doge'],
    ['supports other currencies', '1000 ethereum aud'],
    ['supports coin to coin', '100 doge ark'],
    ['supports vs to coin', '100 usd doge'],
    ['supports vs to vs', '100 usd aud'],
    ['returns an error for invalid amount', 'foo'],
    ['returns an error for invalid coin', '100 asdfasdf'],
    ['returns an error for invalid vs', '100 eth asdfasdf'],
    ['ignores commas in en locale', '400,000.00 doge eur'],
    ['understands de locale', '400.000,00 doge eur', 'de'],
    ['ignores xx locale', '400,000.00 doge eur', 'xx'],
    ['ignores empty locale', '400,000.00 doge eur', ''],
  ],
  '/regret': [
    ['defaults to 10,000 btc sold at $41 usd'],
    ['supports other currencies', '0.5 eth 500 aud'],
    ['congratulates no ragerts', '1000 xrp 1000'],
    ['returns an error for invalid amount', 'foo'],
    ['returns an error for invalid coin', '1 asdfasdf'],
    ['returns an error for invalid soldAt', '1 btc foo'],
    ['returns an error for invalid vs', '1 btc 100 asdfasdf'],
  ],
  '/top10': [['defaults to USD'], ['supports other vs currencies', 'aud']],
  '/top20': [['defaults to USD'], ['supports other vs currencies', 'btc']],
};

const toArray = (t) => (Array.isArray(t) ? t : [t]);

for (const [command, tests] of Object.entries(commandTests)) {
  describe(command, () => {
    for (const [desc, args, locale] of toArray(tests).map(toArray)) {
      it(args ? `${args} - ${desc}` : `- ${desc}`, () => {
        const text = args ? `${command} ${args}` : command;
        return expect(msgReply(text, locale)).resolves.toMatchSnapshot();
      });
    }
  });
}
