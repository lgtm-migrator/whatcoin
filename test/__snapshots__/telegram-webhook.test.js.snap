// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Telegram Webhook /convert () - defaults to 1 btc in usd 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "1 BTC = $38,549",
}
`;

exports[`Telegram Webhook /convert (100, asdfasdf) - returns an error for invalid coin 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "Sorry, I couldn't find asdfasdf. Try using the full name",
}
`;

exports[`Telegram Webhook /convert (100, doge, ark) - supports coin to coin 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "100 DOGE = 2.4112 ARK",
}
`;

exports[`Telegram Webhook /convert (100, eth, asdfasdf) - returns an error for invalid vs 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "Sorry, I couldn't find asdfasdf. Try using the full name",
}
`;

exports[`Telegram Webhook /convert (400,000.00, doge, eur) - ignores commas in en locale 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "400,000 DOGE = €3,205.8",
}
`;

exports[`Telegram Webhook /convert (400,000.00, doge, eur) - ignores empty locale 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "400,000 DOGE = €3,205.8",
}
`;

exports[`Telegram Webhook /convert (400,000.00, doge, eur) - ignores xx locale 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "400,000 DOGE = €3,205.8",
}
`;

exports[`Telegram Webhook /convert (400.000,00, doge, eur) - understands de locale 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "400.000 DOGE = 3.205,8 €",
}
`;

exports[`Telegram Webhook /convert (1000, ethereum, aud) - supports other currencies 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "1,000 ETH = A$1,583,300",
}
`;

exports[`Telegram Webhook /convert (foo) - returns an error for invalid amount 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "Amount 'foo' is not a valid number",
}
`;

exports[`Telegram Webhook /price () - defaults to btc in usd 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*Bitcoin (BTC)* in USD
Current price:  \`$38,549\`
h/d/w/m: \`0.0819%\` / \`11.9%\` / \`33.7%\` / \`101%\` 
Market cap:  \`$718,577,416,969\`
24h volume:  \`$73,907,888,421\`
_(updated Thu, 07 Jan 2021 14:27:08 GMT)_",
}
`;

exports[`Telegram Webhook /price (ETH, GBP) - finds coins by symbol 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*Ethereum (ETH)* in GBP
Current price:  \`£902.26\`
h/d/w/m: \`-0.215%\` / \`6.72%\` / \`63.3%\` / \`103%\` 
Market cap:  \`£103,014,363,730\`
24h volume:  \`£30,758,456,011\`
_(updated Thu, 07 Jan 2021 14:28:44 GMT)_",
}
`;

exports[`Telegram Webhook /price (PEAKDEFI) - finds coins by name 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*PEAKDEFI (PEAK)* in USD
Current price:  \`$0.270058\`
h/d/w/m: \`0.677%\` / \`0.248%\` / \`-1.1%\` / \`4.35%\` 
Market cap:  \`$24,090,107\`
24h volume:  \`$404,893\`
_(updated Thu, 07 Jan 2021 14:23:58 GMT)_",
}
`;

exports[`Telegram Webhook /price (algoha) - finds coins by symbol partial match 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*0.5X Long Algorand Token (ALGOHALF)* in USD
Current price:  \`$8,288.27\`
h/d/w/m: \`?\` / \`?\` / \`?\` / \`?\` 
Market cap:  \`$0\`
24h volume:  \`$53.96\`
_(updated Sat, 21 Nov 2020 14:26:40 GMT)_",
}
`;

exports[`Telegram Webhook /price (asdfasdf, eur) - returns an error for invalid coin 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "Sorry, I couldn't find asdfasdf. Try using the full name",
}
`;

exports[`Telegram Webhook /price (bitcoin, asdasd) - returns an error for invalid vs currency 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "text": "Sorry, I can't list prices in asdasd. Supported base currencies are: btc, eth, ltc, bch, bnb, eos, xrp, xlm, link, dot, yfi, usd, aed, ars, aud, bdt, bhd, bmd, brl, cad, chf, clp, cny, czk, dkk, eur, gbp, hkd, huf, idr, ils, inr, jpy, krw, kwd, lkr, mmk, mxn, myr, ngn, nok, nzd, php, pkr, pln, rub, sar, sek, sgd, thb, try, twd, uah, vef, vnd, zar, xdr, xag, xau",
}
`;

exports[`Telegram Webhook /price (bitm) - finds coins by id partial match 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*BitMoney (BIT)* in USD
Current price:  \`$0.00003131\`
h/d/w/m: \`?\` / \`?\` / \`?\` / \`63.1%\` 
Market cap:  \`$15,947.8\`
24h volume:  \`$4.37\`
_(updated Tue, 05 Jan 2021 14:09:21 GMT)_",
}
`;

exports[`Telegram Webhook /price (doge, bitcoin) - finds vs by coin id 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*Dogecoin (DOGE)* in BTC
Current price:  \`BTC 0.0000002643\`
h/d/w/m: \`-0.419%\` / \`-10.8%\` / \`63.8%\` / \`52.1%\` 
Market cap:  \`BTC 33,766\`
24h volume:  \`BTC 13,775\`
_(updated Thu, 07 Jan 2021 14:27:53 GMT)_",
}
`;

exports[`Telegram Webhook /price (ethereum, eur) - supports other currencies 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*Ethereum (ETH)* in EUR
Current price:  \`€999.21\`
h/d/w/m: \`-0.291%\` / \`7.25%\` / \`63.2%\` / \`104%\` 
Market cap:  \`€114,098,707,361\`
24h volume:  \`€34,063,378,632\`
_(updated Thu, 07 Jan 2021 14:28:44 GMT)_",
}
`;

exports[`Telegram Webhook /price (peakd) - finds coins by name partial match 1`] = `
Object {
  "chat_id": 2,
  "method": "sendMessage",
  "parse_mode": "markdown",
  "text": "*PEAKDEFI (PEAK)* in USD
Current price:  \`$0.270058\`
h/d/w/m: \`0.677%\` / \`0.248%\` / \`-1.1%\` / \`4.35%\` 
Market cap:  \`$24,090,107\`
24h volume:  \`$404,893\`
_(updated Thu, 07 Jan 2021 14:23:58 GMT)_",
}
`;
